### PR TITLE
Printtyp: restore compatibility for compiler-libs users

### PR DESCRIPTION
**Pull request by gasche**
## TL;DR

This PR changes the interface of the Printtyp module to be compatible with its previous type in 5.2 and in all version, un-doing invasive type changes that came with #13169.

### Type changes for printing functions in trunk

in #13169, @Octachron introduced a different Format layer for pretty-printing error messages, which reifies them into an explicit document and gives us hope of more structured communication with the outside world. Most formatting functions in the compiler codebase are used to produce error messages, and those were all switched to the new document type. Some formatting code in the compiler codebase is used for other things, and keeps using Format. Some functions are used in both contexts, and they were changed to use the new document type, with a `compat` layer (translating back into normal Format stuff) when they are used in an old-format context.

In particular, Printtyp (which prints types) is mostly used in error messages, so its functions were all converted to the new types, but sometimes they are used for other things (for example to implement the `-i` option that prints an interface file), and so @Octachron wrote a Printtyp.Compat submodule that contains old-style types for a handful of Printtyp functions used in both contexts -- arguably the most common ones.

### Backward-compatibility issues

In #13311 we started thinking about the amount of breakage in third-party code that uses compiler-libs. I argued that `report_error` functions are commonly used in other people's code (... as soon as they call inside the compiler and do something that may raise an error), and so I proposed to move the type changes to be opt-in rather than opt-out: the interface of the pre-existing functions is restored back to their previous types, and users who want the new type have to use a new function name. This means that third-party compiler-libs users that use `report_error` functions do not need to fix their code.

@Octachron then suggested than Printtyp may benefit from the same treatment: it is another part of the compiler that users commonly access through compiler-libs, see [this github code search](https://github.com/search?q=language%3AOCaml+%22Printtyp.%22&type=code) for example.

### Opt-in type changes for Printtyp

@Octachron's proposal was to move to an opt-in approach to the type change for Printtyp as well, precisely for the functions listed in the `Compat` submodules -- the one that are used in the compiler outside error messages, which hopefully are also the ones more often used in third-party codebases.

When I tried to implement this proposal, I found that the result was painfully inconsistent: there are a lot of printing functions, most of them would have the "new types", say `Printtyp.ident`, but on the other hand just a few of them like `Printtyp.longident` come in two flavors, `Printtyp.longident` with the old type and `Printtyp.longident'` with the new type. This is inconsistent and there is no way for people to guess which functions should fall in which category. (It depends on the history of the compiler codebase, not on another objective criterion.)

So instead I decided to provide both types for *all* printing functions exposed in the module. This PR adds two submodules, Printtyp.Doc (new types) and Printtyp.Compat (old types), and Printtyp.Compat is included in the Printtyp module (so Printtyp.ident is an alias for Printtyp.Compat.indent). As a consequence, the types of most Printtyp.foo functions are back to what they were in 5.2, and one can use `module Printtyp = Printtyp.Doc` to hide those and move to the new types -- this is done often in the parts of the compiler codebase that build error messages.

### CODE REVIEW

**Comment by gasche** [comment]
@Octachron after writing this I seriously considered implementing Printtyp as a functor over an `Fmt` interface. This would require slightly more invasive changes to the implementation, but it would avoid the small code duplication I have for higher-order printers, and conceptually it is a cleaner way to do this. I think that this would be a reasonable thing to do in the future, but I timed out on this work for now, and in any case this need not change the Printtyp *interface* -- providing two concrete instances remains the right choice, I think -- only the implementation, so we can revisit later.

**Comment by Octachron** [comment]
I don't think that such  amount of code duplication to avoid breaking the code of hypothetical users of the internal `Printtyp` API is a good compromise.

In fact, I am taking this proposition as a cry for help for reorganizing the `Printtyp` module in more readable submodules since you are duplicating functions that are not related to printing, and could not find the separation between "internal" functions (the ones which require expert knowledge on `Printtyp` global states) and the "public" ones.

**Comment by gasche** [comment]
I don't understand, what is the code duplication you have in mind? I have the impression that both the .ml and the .mli define each function once, except for one specific case related to functor parameters. In the .mli, this is made in a single signature that is instantiated twice. In the .ml, this is made in the Doc submodule, and then the Compat submodule only expose one-line wrappers/adapters on the Doc functions (no code duplication).

(It may be that the diff is painful to read due to whitespace changes.)

**Comment by Octachron** [comment]
I am speaking of the duplication the interface itself? The `Printers` module type contains functions that are not printers, and internal functions. Duplicating contents with `include`s is still duplicating contents.

**Comment by Octachron** [comment]
Typically, I would rather see for the `Printers` module types:
```ocaml
module type Printers := sig
    type 'a printer
    val label : label_declaration printer
    val ident: Ident.t printer
    val longident : Longident.t printer
    val path: Path.t printer

    val value_description: Ident.t -> value_description printer
    val constructor : constructor_declaration printer
    val type_declaration: Ident.t -> type_declaration printer

    val extension_constructor:
      Ident.t -> extension_constructor printer
      (* Prints extension constructor with the type signature:
           type ('a, 'b) bar += A of float
      *)

    val extension_only_constructor:
      Ident.t -> extension_constructor printer
      (* Prints only extension constructor without type signature:
           A of float
      *)

    val modtype: module_type printer
    val signature: signature printer
    val modtype_declaration: Ident.t -> modtype_declaration printer
    val class_type: class_type printer
    val class_declaration: Ident.t -> class_declaration printer
    val cltype_declaration: Ident.t -> class_type_declaration printer
    val type_expansion :
      type_or_scheme -> Errortrace.expanded_type printer

    (** Print out a type.  This will pick names for type variables, and will not
        reuse names for common type variables shared across multiple type
        expressions.  (It will also reset the printing state, which matters for
        other type formatters such as [prepared_type_expr].)  If you want multiple
        types to use common names for type variables, see [prepare_for_printing] and
        [prepared_type_expr].  *)
    val type_expr: type_expr printer

    val type_scheme: type_expr printer
    val shared_type_scheme: type_expr printer
    (** [shared_type_scheme] is very similar to [type_scheme], but does not reset
        the printing context first.  This is intended to be used in cases where the
        printing should have a particularly wide context, such as documentation
        generators; most use cases, such as error messages, have narrower contexts
        for which [type_scheme] is better suited. *)
end
```
In particular, the `functor_params` module is quite internal and rather tuned to its only-use case in `Includemod_errorprinter`.

**Comment by gasche** [comment]
For me, clearly, everything that takes a `formatter` argument and is exposed in the public API of the module is a printer. If you don't think that something should be exposed as a printer, it should be removed from the public API. (Maybe `functor_params` should be inlined in its caller.)

If we agree with that, I can try to do some extra work to remove a few functions (typically those that produce outcome trees) from the two modules. Note that these functions require zero code duplication, given that I don't even need to rebind them in the Compat module -- they are mentioned in the source exactly once in the .mli and once in the .ml.

But this means that my `module Printtyp = Printtyp.Doc` approach is not going to work anymore, and I have to find another setup that is more pleasant than `module Printtyp = struct include Printtyp include Printtyp.Doc end`, meh. Do you have a suggestion for how you would want to do this?

**Comment by gasche** [comment]
@Octachron and myself have different views on this PR:
- I would be in favor of merging it first for backward-compatibility (before 5.3), and *then* improving the Printtyp interface (which I see as an independent endeavor)
- @Octachron does not like the proposed interface; he volunteers to try to propose an alternative PR before 5.3 (for example by the end of July), but is not sold on the idea of merging this one if he doesn't have the time to propose another. More discussion needed in that case.

**Comment by gasche** [comment]
Superseded by #13336 (not merged yet, but hopefully soon).
